06-security.md

```diff
@@ -1 +1,426 @@
+
+#### Limit the number of Login attempts per unit of time.
+
+If the login form is accessible from the Internet, the number of (failed) login attempts must be limited per unit of time to prevent brute forcing. The counting must be done in the user account, not in the session (as a new session may be used for each attempt). 
+
+#### Consider using password complexity requirements and blacklists
+
+While having a these will not prevent [pattern based brute forcing techniques](https://www.youtube.com/watch?v=zUM7i8fsf0g) it will prevent users from picking too simple or common passwords.
+Good blacklists can be found [here](https://github.com/danielmiessler/SecLists/tree/master/Passwords). 
+
+### Configure Cross Origin Resource Sharing if needed
+
+If a browser application that resides on a different domain access needs to access the API the single origin policy (SOP) will prevent the browser from reading the responses. There is a way around this by supporting the OPTIONS request method and returning appropriate CORS headers. Before performing a request on a different domain the browser will first issue an OPTIONS request to see if it is allowed to perform the request. Typically a CORS response should look something like:
+
+```
+Access-Control-Allow-Credentials: true
+Access-Control-Allow-Origin: https://somedomain.org, https://otherdomain.org
+Access-Control-Allow-Methods: GET, OPTIONS, POST, PUT, PATCH, DELETE
+Access-Control-Allow-Headers: Origin, X-Requested-With, Content-Type, Accept, X-XSRF-TOKEN
+Access-Control-Max-Age: 3600
+```
+
+Access-Control-Allow-Credentials allows the passing of a session cookie. 
+The Access-Control-Allow-Origin lists the domains that are allowed to access the API.
+The Access-Control-Allow-Methods list the permissible HTTP Methods.
+Access-Control-Allow-Headers informs the browser of the allowed headers to send.
+Finally Access-Control-Max-Age informs the browser on how long the information in the response may be cached.
+
+*Do not* use the wildcard on Access-Control-Allow-Origin unless the API must be accessible to the world.  
+```
+Access-Control-Allow-Origin: *
+``` 
+
+### Use browser security response headers.
+
+Various [headers](https://www.owasp.org/index.php/List_of_useful_HTTP_headers) exist to prevent abuse of resources in and from the browser. 
+
+```
+X-Frame-Options: deny
+X-XSS-Protection: 1; mode=block
+X-Content-Type-Options: nosniff
+```
+
+The `X-Frame-Options` header disallows the contents to be displayed in an IFrame. The `X-XSS-Protection` header makes the browser apply a filter to the URL to prevent reflected Cross Site Scripting. Finally, the `X-Content-Type-Options` header disallows the browser from guessing the content type. 
+
+### Disable caching
+
+In general responses from Rest API's should not be cached. This is partly because the data returned may be sensitive and also because the data must be fresh (so not an old copy of the data). For this various headers must be set: 
```

Comment:
Maybe emphasize the word `general` so that people do not whine: "but... what if..."